### PR TITLE
feat: new `mongodbatlas_stream_instances` data source

### PR DESCRIPTION
**Pull request by AgustinBettati**
## Description

Includes:

- Implementation of mongodbatlas_stream_instance data source
- Acceptance tests
- Documentation
- Definition of common pagination schema as proposed in https://github.com/mongodb/terraform-provider-mongodbatlas/pull/1693

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [x] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR.
- [ ] This change requires a documentation update
- [ ] Documentation fix/enhancement

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have read the [contribution guidelines](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/CONTRIBUTING.md)
- [x] I have checked that this change does not generate any credentials and that **they are NOT accidentally logged anywhere**.
- [x] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run make fmt and formatted my code
- [x] If changes include deprecations or removals, I defined an isolated PR with a relevant title as it will be used in the auto-generated changelog.

### CODE REVIEW

File: .github/workflows/acceptance-tests.yml [hunk 1]
```diff
@@ -59,7 +59,7 @@ jobs:
             - 'internal/service/cluster/*.go'
           search_deployment:
             - 'internal/service/searchdeployment/*.go'
-          stream_instance:
+          stream:
```
Comment (AgustinBettati) [inline .github/workflows/acceptance-tests.yml:68]: transitioned to a `stream` CI group to include the coming stream_connection resource in here as well. Both define acceptance tests that run fast.

Comment (lantoli) [reply .github/workflows/acceptance-tests.yml:68]: good

File: internal/service/streaminstance/model_stream_instance.go [hunk 2]
```diff
@@ -6,6 +6,8 @@ import (
 	"github.com/hashicorp/terraform-plugin-framework/diag"
 	"github.com/hashicorp/terraform-plugin-framework/types"
 	"github.com/hashicorp/terraform-plugin-framework/types/basetypes"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/id"
```
Comment (lantoli) [inline internal/service/streaminstance/model_stream_instance.go:9]: does it need v2 being a new resource?

**Review by marcosuma — APPROVED** [review]

File: internal/service/streaminstance/data_source_stream_instance.go [hunk 3]
```diff
@@ -26,31 +26,39 @@ type streamInstanceDS struct {
...
-					},
-					"region": schema.StringAttribute{
-						Computed: true,
-					},
+		Attributes: DSAttributes(true),
+	}
+}
+
+// DSAttributes returns the attribute definitions for a single stream instance.
+// `withArguments` is used to mark certain attributes as required (for singular data source) or as computed (for plural data source)
```
Comment (kanchana-mongodb) [inline internal/service/streaminstance/data_source_stream_instance.go]:
```suggestion
// `withArguments` marks certain attributes as required (for singular data source) or as computed (for plural data source)
```

File: website/docs/d/stream_instances.html.markdown [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+* `project_id` - (Required) Unique 24-hexadecimal digit string that identifies your project.
+
+* `page_num` - (Optional) Number of the page that displays the current set of the total objects that the response returns. Defaults to `1`.
+* `items_per_page` - (Optional) Number of items that the response returns per page, up to a maximum of `500`. Defaults to `100`.
+
+
+## Attributes Reference
+
+In addition to all arguments above, the following attributes are exported:
```
Comment (kanchana-mongodb) [inline website/docs/d/stream_instances.html.markdown]:
```suggestion
In addition to all arguments above, it also exports the following attributes:
```

File: website/docs/d/stream_instances.html.markdown [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+* `results` - A list where each element contains a Stream Instance.
+* `total_count` - Count of the total number of items in the result set. It may be greater than the number of objects in the results array if the entire result set is paginated.
```
Comment (kanchana-mongodb) [inline website/docs/d/stream_instances.html.markdown]:
```suggestion
* `total_count` - Count of the total number of items in the result set. The count might be greater than the number of objects in the results array if the entire result set is paginated.
```

File: website/docs/d/stream_instances.html.markdown [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+* `instance_name` - Human-readable label that identifies the stream instance.
+* `data_process_region` - Defines the cloud service provider and region where MongoDB Cloud performs stream processing. See [data process region](#data-process-region).
+* `hostnames` - List that contains the hostnames assigned to the stream instance.
+
+### Data Process Region
+
+* `cloud_provider` - Label that identifies the cloud service provider where MongoDB Cloud performs stream processing. The [MongoDB Atlas API](https://www.mongodb.com/docs/atlas/reference/api-resources-spec/#tag/Streams/operation/createStreamInstance) describes the valid values.
+* `region` - Name of the cloud provider region hosting Atlas Stream Processing. The [MongoDB Atlas API](https://www.mongodb.com/docs/atlas/reference/api-resources-spec/#tag/Streams/operation/createStreamInstance) describes the valid values.
+  
+For more information see: [MongoDB Atlas API - Stream Instance](https://www.mongodb.com/docs/atlas/reference/api-resources-spec/#tag/Streams/operation/createStreamInstance) Documentation.
```
Comment (kanchana-mongodb) [inline website/docs/d/stream_instances.html.markdown]:
```suggestion
To learn more, see: [MongoDB Atlas API - Stream Instance](https://www.mongodb.com/docs/atlas/reference/api-resources-spec/#tag/Streams/operation/createStreamInstance) Documentation.
```

**Review by kanchana-mongodb — APPROVED** [review]
Couple of nits, LGTM.

File: internal/service/streaminstance/model_stream_instance.go (hunk 2 above)
Comment (AgustinBettati) [reply internal/service/streaminstance/model_stream_instance.go:9]: This is a helper function provided by hashicorp is for generating a unique id, did not find a framework alternative but nevertheless we are not coupling to sdk v2 logic, this is simply for a obtaining a random id.

**Pull request merged**